experimental/inductor.py

```diff
@@ -0,0 +1,669 @@
+        #      graph to inductor. Inductor expects only tensors as inputs
+        #    - This works since all the non-tensor inputs should be unused at this point,
+        #      since in principle we should have converted all getattrs into inputs.
+        #    - We keep the indices of unused inputs in "unused_inputs_idx"
+        unused_inputs = []
+        unused_inputs_idx = []
+
+        for i, p in enumerate(k2.graph.nodes):
+            if p.op != 'placeholder':
+                break
+            if len(p.users) == 0:
+                unused_inputs.append(p)
+                unused_inputs_idx.append(i)
+
+        for p in unused_inputs:
+            k2.graph.erase_node(p)
+
+        k2.recompile()
+        print_graph('before_functionalize', k2)
+ 
+        # 4. call make_fx a second time with functionalization
+        with torch.no_grad():
+            used_args_and_attrs = prepare_args(attr_vals, unused_inputs_idx, args)
+            k3 = make_fx(functionalize(k2), decomposition_table=all_decomps)
+            k4 = k3(*used_args_and_attrs, **kwargs)
+
+        # 5. remove useless "alias" nodes
+        for p in k4.graph.nodes:
+            if p.target == torch.ops.aten.alias.default:
+                p.replace_all_uses_with(p.args[0])
+
+        # 6. functionalization adds getattrs sometimes, so move them to inputs again
+        #    (hoping they'll be deterministic constants...)
+        attr_vals += convert_getattrs_to_inputs(k4)
+        print_value('attr_vals_after_functionalize', attr_vals)
+
+        print_graph('after functionalize before DCE', k4)
+
+        # 7. eliminate dead code but make sure we're still keeping the final copy_
+        torch.fx.node._side_effectful_functions.add(torch.ops.aten.copy_.default)
+        k4.graph.eliminate_dead_code()
+        k4.graph.lint()
+        k4.recompile()
+
+        print_graph('after functionalize after DCE', k4)
+
+        return k4, attr_vals, unused_inputs_idx
+
+    return call
+
+DEFAULT_SCHEMA = 'default' 
+
+
+def track_state_dict_materialization(map: Dict[FakeTensor, Tensor], fake_state_dict, real_state_dict):
+    for k, v in fake_state_dict.items():
+        real_k = map[k] if isinstance(k, Tensor) else k
+        if isinstance(v, Tensor):
+            map[v] = real_state_dict[real_k]
+        elif isinstance(v, dict):
+            track_state_dict_materialization(map, v, real_state_dict[real_k])
+
+
+def warmup_optimizer(optim):
+    # unfortunatelly, in order to initialize the optimizer states
+    # we need to call optimizer.step(). In order to do that, we will initialize
+    # parameter gradients with zero tensors just so we can run a step.
+    for group in optim.param_groups:
+        for param in group['params']:
+            param.grad = torch.zeros_like(param)
+    # run an optimizer step just to initialize the optimizer states
+    # multiple issues here:
+    # 1. this will make the optimizer think we're at the 2nd iteration
```

Comment:
if we do `make_inductor_fn` directly on the train_step, would we still need to warm up optimizer? because the train_step itself contains foward + backward + optim.step so it should be warmed up already during the first time we `make_fx` it?

Reply:
The first time we make_fx it, it will be on fake mode. This is required because we're initially tracing with "global shapes", which would make our device OOM if we were not in fake mode.

Then, the second time we trace it, we still didn't do SPMD expansion, so even though we could, in theory, run the train_step with eager DTensor, I thought we would prefer to avoid that, since we may want to do SPMD expansion using more sophisticated propagation techniques in the future. But it would be an option for sure.